<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Introduction to R: <code>Bioconductor</code> and other bioinformatics packages</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../intro-bioconductor.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Introduction to R
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Introduction to R
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to R
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 57%" class="percentage">
    57%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 57%" aria-valuenow="57" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../intro-bioconductor.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction to R and Rstudio</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="reading-in-data-readr-and-readxl.html">2. Reading in data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="tidying-data.html">3. Tidy data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="manipulating-data.html">4. Data wrangling</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="visualizing-data.html">5. Data visualization</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        <code>Bioconductor</code> and other bioinformatics packages
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction-to-bioconductor">Introduction to Bioconductor</a></li>
<li><a href="#biostrings">Biostrings</a></li>
<li><a href="#plotting-genomic-information">Plotting genomic information</a></li>
<li><a href="#rna-seq">RNA-seq</a></li>
<li><a href="#resources">Resources</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="intro-rmarkdown.html">7. Using <code>R markdown</code> for notebooks and documents</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="git.html">8. Version control with <code>git</code></a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="worked-example.html">9. Case study</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/visualizing-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/intro-rmarkdown.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/visualizing-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Data visualization</a>
        <a class="chapter-link float-end" href="../instructor/intro-rmarkdown.html" rel="next">Next: Using &lt;code&gt;R markdown&lt;/code&gt;... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1><code>Bioconductor</code> and other bioinformatics packages</h1>
        <p> Last updated on 2022-12-20 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/intro-bioconductor.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 22 minutes </p>
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What is bioconductor?</li>
<li>What are some bioconductor packages?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Install bioconductor packages with <code>BiocManager</code>
</li>
<li>Be aware of some <code>Bioconductor packages</code>
</li>
</ul></div>
</div>
</div>
</div>
</div>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="do-i-need-to-do-this-lesson" class="callout-inner">
<h3 class="callout-title">Do I need to do this lesson?<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>If you can answer the following questions, you can skip this lesson.</p>
<ol style="list-style-type: decimal"><li>How do you install <code>Bioconductor</code> packages?</li>
<li>How can you get help when using <code>Bioconductor</code> packages?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li>Use <code>BiocManager::install()</code>.</li>
<li>In the same way as other <code>R</code> packages. All will have a <code>vignette</code> demonstrating how to use the package, and most will have some other guides demonstrating aspects of the package.</li>
</ol></div>
</div>
</div>
</div>
<figure><img src="http://bioconductor.org/images/logo_bioconductor.gif" alt="" class="figure mx-auto d-block"><figcaption>Bioconductor logo</figcaption></figure><section id="introduction-to-bioconductor"><h2 class="section-heading">Introduction to Bioconductor<a class="anchor" aria-label="anchor" href="#introduction-to-bioconductor"></a>
</h2>
<hr class="half-width"><p>We’ll quickly take a look at the<code>Bioconductor</code> ecosystem, a collection of <code>R</code> packages for bioinformatics. Usually these solve specific problems (like RNA-seq analysis or plotting genomic data), so they won’t be as applicable to everyone here.</p>
<p>The goal of this lesson is therefore to let you know about some packages from <code>bioconductor</code> that I’ve found useful, but we won’t actually go into any detail about how to use them. All of the <code>Bioconductor</code> packages have vignettes that will show you how to use the core functionality of the package.</p>
<div class="section level3">
<h3 id="installing-packages-with-biocmanager">Installing packages with <code>BiocManager</code>
<a class="anchor" aria-label="anchor" href="#installing-packages-with-biocmanager"></a></h3>
<p>While typically you’d use <code>install.packages()</code> to install <code>R</code> packages, <code>Bioconductor</code> has it’s own package manager. The package manager is called <a href="https://cran.r-project.org/web/packages/BiocManager/vignettes/BiocManager.html" class="external-link"><code>BiocManager</code></a>, and you can install it as you would any other package:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install BiocManager</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<p>You can then use this package to install other <code>Bioconductor</code> packages, for example:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span></code></pre>
</div>
<p><code>Bioconductor</code> also has it’s own release schedule, which are designed to work with particular versions of <code>R</code>.</p>
<p>At the time of writing, the <a href="https://bioconductor.org/install/" class="external-link">current version is 3.16</a>, which works with <code>R</code> version 4.2.0. If you have an older version of <code>R</code>, you will need to use an older version of <code>Bioconductor</code> as well - there’s <a href="https://bioconductor.org/about/release-announcements/" class="external-link">a list of the older releases on the Bioconductor website</a>. You can tell <code>BiocManager</code> which release to use:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version<span class="op">=</span><span class="st">"3.16"</span><span class="op">)</span></span></code></pre>
</div>
<p>This won’t actually install anything, but all future installs will use version 3.16.</p>
</div>
</section><section id="biostrings"><h2 class="section-heading">Biostrings<a class="anchor" aria-label="anchor" href="#biostrings"></a>
</h2>
<hr class="half-width"><p>A common task (at least for me) is working with biological sequences (e.g. DNA, RNA, amino acid). A useful package for this is <a href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html" class="external-link"><code>Biostrings</code></a>.</p>
<p>Install this package with the following code</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span></span></code></pre>
</div>
<p>Biostrings contains useful variable and functions for working with sequences. For example, you can translate a nucleotide sequence:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load library</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">biostrings</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create DNAString object</span></span>
<span><span class="va">my_seq</span> <span class="op">&lt;-</span> <span class="fu">DNAString</span><span class="op">(</span><span class="st">"GTAAGTGTATGC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># translate to amino acids</span></span>
<span><span class="fu">translate</span><span class="op">(</span><span class="va">my_seq</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>4-letter AAString object
seq: VSVC</code></pre>
</div>
<p>Or calculate the frequency of each nucleotide:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">alphabetFrequency</span><span class="op">(</span><span class="va">my_seq</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>A C G T M R W S Y K V H D B N - + . 
3 1 4 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 </code></pre>
</div>
<p>Or the codon frequencies</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">trinucleotideFrequency</span><span class="op">(</span><span class="va">my_seq</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>AAA AAC AAG AAT ACA ACC ACG ACT AGA AGC AGG AGT ATA ATC ATG ATT CAA CAC CAG CAT 
  0   0   1   0   0   0   0   0   0   0   0   1   0   0   1   0   0   0   0   0 
CCA CCC CCG CCT CGA CGC CGG CGT CTA CTC CTG CTT GAA GAC GAG GAT GCA GCC GCG GCT 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
GGA GGC GGG GGT GTA GTC GTG GTT TAA TAC TAG TAT TCA TCC TCG TCT TGA TGC TGG TGT 
  0   0   0   0   2   0   1   0   1   0   0   1   0   0   0   0   0   1   0   1 
TTA TTC TTG TTT 
  0   0   0   0 </code></pre>
</div>
<p>It also contains a codon table</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GENETIC_CODE</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>TTT TTC TTA TTG TCT TCC TCA TCG TAT TAC TAA TAG TGT TGC TGA TGG CTT CTC CTA CTG 
"F" "F" "L" "L" "S" "S" "S" "S" "Y" "Y" "*" "*" "C" "C" "*" "W" "L" "L" "L" "L" 
CCT CCC CCA CCG CAT CAC CAA CAG CGT CGC CGA CGG ATT ATC ATA ATG ACT ACC ACA ACG 
"P" "P" "P" "P" "H" "H" "Q" "Q" "R" "R" "R" "R" "I" "I" "I" "M" "T" "T" "T" "T" 
AAT AAC AAA AAG AGT AGC AGA AGG GTT GTC GTA GTG GCT GCC GCA GCG GAT GAC GAA GAG 
"N" "N" "K" "K" "S" "S" "R" "R" "V" "V" "V" "V" "A" "A" "A" "A" "D" "D" "E" "E" 
GGT GGC GGA GGG 
"G" "G" "G" "G" 
attr(,"alt_init_codons")
[1] "TTG" "CTG"</code></pre>
</div>
<p>It has many more features which you can explore if you need to use this package.</p>
</section><section id="plotting-genomic-information"><h2 class="section-heading">Plotting genomic information<a class="anchor" aria-label="anchor" href="#plotting-genomic-information"></a>
</h2>
<hr class="half-width"><p>Another relatively common task for me is plotting genomic information. There are a number of ways to do this, but here I’ll just give one example - plotting the density of SNPs in the human albumin gene (ALB).</p>
<p>First, I grabbed the <a href="https://gnomad.broadinstitute.org/gene/ENSG00000163631?dataset=gnomad_r3" class="external-link">SNP information for this gene from <code>gnomAD</code></a> as a csv, and read in this information using <code>readr::read_csv()</code></p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load data</span></span>
<span><span class="va">alb</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu">read_csv</span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu">here</span><span class="op">(</span><span class="st">"episodes"</span>, <span class="st">"data"</span>, <span class="st">"gnomAD_v3.1.2_ENSG00000163631_2022_11_25_11_56_32.csv"</span><span class="op">)</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<p>Next, we need to convert the SNP locations into a <code>GRanges</code> object - this is an object from the <code>GenomicRanges</code> package that is designed to store genomic coordinates and associated metadata. I won’t go into detail about this either, but you can find more information about this package <a href="https://bioconductor.org/packages/devel/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html" class="external-link">in the documentation</a>.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for pipe</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">magrittr</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'magrittr'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:GenomicRanges':

    subtract</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alb_gr</span> <span class="op">&lt;-</span> <span class="va">alb</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="co"># add "chr" to chromosome number</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>Chromosome <span class="op">=</span> <span class="fu">paste0</span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">Chromosome</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="co"># convert to GRanges object</span></span>
<span>  <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu">makeGRangesFromDataFrame</span><span class="op">(</span></span>
<span>    seqnames.field<span class="op">=</span><span class="st">"Chromosome"</span>,</span>
<span>    start.field <span class="op">=</span> <span class="st">"Position"</span>,</span>
<span>    end.field<span class="op">=</span><span class="st">"Position"</span>,</span>
<span>    keep.extra.columns <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">alb_gr</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 850 ranges and 0 metadata columns:
        seqnames    ranges strand
           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
    [1]     chr4  73397115      *
    [2]     chr4  73397120      *
    [3]     chr4  73397122      *
    [4]     chr4  73397122      *
    [5]     chr4  73397123      *
    ...      ...       ...    ...
  [846]     chr4  73421175      *
  [847]     chr4  73421187      *
  [848]     chr4  73421188      *
  [849]     chr4  73421189      *
  [850]     chr4  73421193      *
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<p>Now we have the data loaded, we’re ready to start making the plot.</p>
<div class="section level3">
<h3 id="karyoploter">KaryoploteR<a class="anchor" aria-label="anchor" href="#karyoploter"></a></h3>
<p>Here I’m more or less following part of the <a href="https://bernatgel.github.io/karyoploter_tutorial//Examples/EncodeEpigenetics/EncodeEpigenetics.html" class="external-link">karyploteR tutorial</a>.</p>
<p>This is human data, so we plot the albumin locus in the human genome. I used the <a href="https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&amp;lastVirtModeType=default&amp;lastVirtModeExtraState=&amp;virtModeType=default&amp;virtMode=0&amp;nonVirtPosition=&amp;position=chr4%3A73402360%2D73423322&amp;hgsid=1504808111_WcWVaHMGon8yjsLsfnwXMp51XDT0" class="external-link">UCSC genome brower</a> to identify the coordinates for ALB (roughly chr4:73,394,000-73,425,000), but you could equally use <a href="https://asia.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163631;r=4:73397114-73421482" class="external-link">Ensembl</a>.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">karyoploteR</span><span class="op">)</span></span>
<span><span class="co"># this package contains transcript information for the human genome</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># region to plot</span></span>
<span><span class="va">region</span> <span class="op">&lt;-</span> <span class="st">"chr4:73,394,000-73,425,000"</span></span>
<span></span>
<span><span class="co"># draw axis</span></span>
<span><span class="va">kp</span> <span class="op">&lt;-</span> <span class="fu">plotKaryotype</span><span class="op">(</span>zoom <span class="op">=</span> <span class="va">region</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># transcript information</span></span>
<span><span class="va">genes.data</span> <span class="op">&lt;-</span> <span class="fu">makeGenesDataFromTxDb</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span>,</span>
<span>                                    karyoplot<span class="op">=</span><span class="va">kp</span>,</span>
<span>                                    plot.transcripts <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                    plot.transcripts.structure <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  1662 genes were dropped because they have exons located on both strands
  of the same reference sequence or on more than one reference sequence,
  so cannot be represented by a single genomic range.
  Use 'single.strand.genes.only=FALSE' to get all the genes in a
  GRangesList object, or use suppressMessages() to suppress this message.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add gene names</span></span>
<span><span class="va">genes.data</span> <span class="op">&lt;-</span> <span class="fu">addGeneNames</span><span class="op">(</span><span class="va">genes.data</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: org.Hs.eg.db</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># merge overlapping transcripts</span></span>
<span><span class="va">genes.data</span> <span class="op">&lt;-</span> <span class="fu">mergeTranscripts</span><span class="op">(</span><span class="va">genes.data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># draw genes</span></span>
<span><span class="fu">kpPlotGenes</span><span class="op">(</span><span class="va">kp</span>, data<span class="op">=</span><span class="va">genes.data</span>, r0<span class="op">=</span><span class="fl">0</span>, r1<span class="op">=</span><span class="fl">0.15</span><span class="op">)</span></span>
<span><span class="co"># draw SNP density</span></span>
<span><span class="fu">kpPlotDensity</span><span class="op">(</span><span class="va">kp</span>, data<span class="op">=</span><span class="va">alb_gr</span>, window<span class="op">=</span><span class="fl">100</span>, r0<span class="op">=</span><span class="fl">0.3</span>, r1<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/intro-bioconductor-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level3">
<h3 id="circlize">circlize<a class="anchor" aria-label="anchor" href="#circlize"></a></h3>
<p>Another package for plotting genomic information is <code>circlize</code>, which plots chromosome in a circle rather than in a linear format. I won’t cover this in any detail, but you can check out the <a href="https://jokergoo.github.io/circlize_book/book/" class="external-link">documentation</a> for more information.</p>
</div>
</section><section id="rna-seq"><h2 class="section-heading">RNA-seq<a class="anchor" aria-label="anchor" href="#rna-seq"></a>
</h2>
<hr class="half-width"><p>One common tasks people use <code>Bioconductor</code> packages for is RNA-seq analysis. The steps in RNA-seq analysis are roughly the following:</p>
<ol style="list-style-type: decimal"><li>Alignment of reads using splice-aware mapper (I use <a href="https://github.com/alexdobin/STAR" class="external-link"><code>STAR</code></a>)</li>
<li>Counting reads aligned to each exon/transcript/gene (I use <a href="https://htseq.readthedocs.io/en/master/htseqcount.html" class="external-link"><code>htseq-count</code> from <code>HTSeq</code></a>)</li>
<li>Downstream analysis with counts - batch correction, differential expression, gene ontology, pathway analysis, etc</li>
</ol><p>While it’s possible to do the entire RNA-seq analysis pipeline, from reads to differential expressed genes, in <code>R</code>, you’ll notice that I usually do the first two steps using other software. This mostly comes down to momentum - this is how I was taught - but also has to do with the perception that <code>R</code> easier to use but slower and more memory-intensive than other languages (like <code>C++</code>, the language in which <code>STAR</code> is written).</p>
<p>We could spend a whole session (or workshop) on RNA-seq, so instead of trying to cover all of this material, I’ll instead link to several resources that might be helpful for experimental design and data processing.</p>
<ul><li>
<a href="https://www.melbournebioinformatics.org.au/tutorials/tutorials/rna_seq_exp_design/rna_seq_experimental_design/" class="external-link">Notes on RNA-seq experimental design</a> (Very important!! Without proper experimental design you are only wasting time and money.)</li>
<li>The <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" class="external-link">DESeq2 vignette</a>
</li>
<li>The <a href="http://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeR.pdf" class="external-link">EdgeR vignette</a>
</li>
<li><a href="https://combine-australia.github.io/2018-09-26-RNAseq-Melbourne/" class="external-link">Materials from a two-day RNA-seq workshop</a></li>
<li><a href="https://hbctraining.github.io/DGE_workshop_salmon/lessons/07_DGE_summarizing_workflow.html" class="external-link">Sample differential gene expression workflow</a></li>
</ul><p>If you’re more interested in single cell analysis, the workflow is in principle the same, but there is the added complication of counting barcodes from each cell. If your data is from a 10X workflow, the easiest way to go from reads to counts is using their <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_in" class="external-link"><code>cellranger</code> pipeline</a>. Once you have the count matrices, you can then use <code>R</code> packages for further analysis.</p>
<ul><li><a href="https://satijalab.org/seurat/" class="external-link">Seurat</a></li>
<li><a href="https://cole-trapnell-lab.github.io/monocle3/" class="external-link">Monocle3</a></li>
</ul></section><section id="resources"><h2 class="section-heading">Resources<a class="anchor" aria-label="anchor" href="#resources"></a>
</h2>
<hr class="half-width"><ul><li><a href="https://cran.r-project.org/web/packages/BiocManager/vignettes/BiocManager.html" class="external-link">BiocManager vignette</a></li>
<li><a href="https://web.stanford.edu/class/bios221/labs/biostrings/lab_1_biostrings.html" class="external-link">Biostrings lab</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html" class="external-link">Biostrings documentation</a></li>
<li><a href="https://bernatgel.github.io/karyoploter_tutorial/" class="external-link">karyoploteR documentation</a></li>
<li><a href="https://jokergoo.github.io/circlize_book/book/" class="external-link">circlize documentation</a></li>
<li>
<a href="https://www.melbournebioinformatics.org.au/tutorials/tutorials/rna_seq_exp_design/rna_seq_experimental_design/" class="external-link">Notes on RNA-seq experimental design</a> (Very important!! Without proper experimental design you are only wasting time and money.)</li>
<li>The <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" class="external-link">DESeq2 vignette</a>
</li>
<li>The <a href="http://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeR.pdf" class="external-link">EdgeR vignette</a>
</li>
<li><a href="https://combine-australia.github.io/2018-09-26-RNAseq-Melbourne/" class="external-link">Materials from a two-day RNA-seq workshop</a></li>
<li><a href="https://hbctraining.github.io/DGE_workshop_salmon/lessons/07_DGE_summarizing_workflow.html" class="external-link">Sample differential gene expression workflow</a></li>
<li><a href="https://satijalab.org/seurat/" class="external-link">Seurat documentation</a></li>
<li><a href="https://cole-trapnell-lab.github.io/monocle3/" class="external-link">Monocle3 documentation</a></li>
</ul><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Use <code>BiocManager</code> to install <code>Bioconductor</code> packages</li>
<li>Use <code>Biostrings</code> for working with biological sequences</li>
<li>Use <code>karyoploteR</code> or <code>circlize</code> for plotting genomic features</li>
<li>Use <code>EdgeR</code> and <code>DEseq2</code> for differential expression analysis</li>
<li>Use <code>Seurat</code> and <code>Monocle3</code> for single-cell RNA-seq data analysis</li>
</ul></div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/visualizing-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/intro-rmarkdown.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/visualizing-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Data visualization</a>
        <a class="chapter-link float-end" href="../instructor/intro-rmarkdown.html" rel="next">Next: Using &lt;code&gt;R markdown&lt;/code&gt;... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/intro-bioconductor.Rmd" class="external-link">Edit on GitHub</a>
        
        
        
        | <a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries/workbench-template-rmd/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:sscott@cmri.org.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.11.3" class="external-link">sandpaper (0.11.3)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.3.2" class="external-link">pegboard (0.3.2)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.11" class="external-link">varnish (0.2.11)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries.github.io/workbench-template-rmd/instructor/intro-bioconductor.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "&lt;code&gt;Bioconductor&lt;/code&gt; and other bioinformatics packages",
  "creativeWorkStatus": "active",
  "url": "https://carpentries.github.io/workbench-template-rmd/instructor/intro-bioconductor.html",
  "identifier": "https://carpentries.github.io/workbench-template-rmd/instructor/intro-bioconductor.html",
  "dateCreated": "2022-12-08",
  "dateModified": "2022-12-20",
  "datePublished": "2022-12-20"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

